File: typescript/emit.ts
/**
 * Emit attestation artifacts from the shared vectors, for the cross-language interop check.
 */
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import type { JWK } from "jose";

import {
  ClientAttestationBuilder,
  ClientAttestationCredential,
  SigningKeyPair,
} from "./src/index.js";

interface Inputs {
  attester: { iss: string; alg: string; jwk: JWK };
  instance: { alg: string; jwk: JWK };
  client_id: string;
  audience: string;
  token_endpoint: string;
  attestation_ttl_seconds: number;
}

const HERE = dirname(fileURLToPath(import.meta.url));
const VECTORS = resolve(HERE, "..", "vectors");

const inp: Inputs = JSON.parse(readFileSync(join(VECTORS, "inputs.json"), "utf8"));

const attester = await SigningKeyPair.fromJwk(inp.attester.jwk, inp.attester.alg);
const instance = await SigningKeyPair.fromJwk(inp.instance.jwk, inp.instance.alg);

const attestation = await new ClientAttestationBuilder(attester, inp.attester.iss)
  .clientId(inp.client_id)
  .confirmationKey(instance)
  .expiresIn(inp.attestation_ttl_seconds)
  .build();

const cred = new ClientAttestationCredential(attestation, instance);
const pop = await cred.popHeaders(inp.client_id, inp.audience);
const dpop = await cred.dpopHeaders("POST", inp.token_endpoint);

const out = {
  language: "typescript",
  attestation,
  pop: pop["OAuth-Client-Attestation-PoP"],
  dpop: dpop["DPoP"],
  audience: inp.audience,
  tokenEndpoint: inp.token_endpoint,
  clientId: inp.client_id,
};

mkdirSync(join(VECTORS, "out"), { recursive: true });
writeFileSync(join(VECTORS, "out", "typescript.json"), JSON.stringify(out, null, 2));
console.log("wrote vectors/out/typescript.json");

File: python/emit.py
"""Emit attestation artifacts from the shared vectors, for the cross-language interop check."""
import json
import os

from client_attestation_sdk import ClientAttestationBuilder, ClientAttestationCredential, SigningKeyPair

HERE = os.path.dirname(os.path.abspath(__file__))
VECTORS = os.path.normpath(os.path.join(HERE, "..", "vectors"))

with open(os.path.join(VECTORS, "inputs.json")) as fh:
    inp = json.load(fh)

attester = SigningKeyPair.from_jwk(inp["attester"]["jwk"], inp["attester"]["alg"])
instance = SigningKeyPair.from_jwk(inp["instance"]["jwk"], inp["instance"]["alg"])

attestation = (
    ClientAttestationBuilder(attester, inp["attester"]["iss"])
    .client_id(inp["client_id"])
    .confirmation_key(instance)
    .expires_in(inp["attestation_ttl_seconds"])
    .build()
)

cred = ClientAttestationCredential(attestation, instance)
pop = cred.pop_headers(inp["client_id"], inp["audience"])
dpop = cred.dpop_headers("POST", inp["token_endpoint"])

out = {
    "language": "python",
    "attestation": attestation,
    "pop": pop["OAuth-Client-Attestation-PoP"],
    "dpop": dpop["DPoP"],
    "audience": inp["audience"],
    "tokenEndpoint": inp["token_endpoint"],
    "clientId": inp["client_id"],
}

os.makedirs(os.path.join(VECTORS, "out"), exist_ok=True)
with open(os.path.join(VECTORS, "out", "python.json"), "w") as fh:
    json.dump(out, fh, indent=2)
print("wrote vectors/out/python.json")
